### ISSUE TITLE
DbContextExtensions does not use _deletedentities collection.

### ISSUE THREAD

**Issue by harrisunderwork** [tonysneed/trackable-entities#50]
Currently, when I remove an item from collection and call ApplyChanges(), it only traverses on items which are available, but _deletedentities contain removed entries, which need to be attached to context so that objectcontextmanager can give deleted entries too.

It was impacting this code :

var changes = manager.GetObjectStateEntries(EntityState.Added | EntityState.Modified | EntityState.Deleted);

Regards.

**Comment by tonysneed** [comment]
Are you calling GetChanges first?

**Comment by harrisunderwork** [comment]
Nopes.

**Comment by tonysneed** [comment]
I think what you're going to want to do is write code on the server which would ordinarily go on the client, which means calling GetChanges first, to obtain adds, updates and deletes. Then pass those to ApplyChanges. Have a look at the VS2013 samples, the client app, for an example of how that's done.

**Comment by harrisunderwork** [comment]
Tony,

I am not using GetChanges() method at all. Its because I am using ESB
pattern, so I cant send only changed part as I may need to modify unchanged
entities too.

Problem lies with ApplyChanges() method. I think it is tracking everything
except deleted entities. As deleted items don't exist in collection, it is
unable to attach it to contextobject.

Regards.

**Comment by tonysneed** [comment]
I think the most straightforward would be to mark the items as deleted rather than removing them from the collection. Then call ApplyChanges and SaveChanges, which will in fact delete them.

**Comment by harrisunderwork** [comment]
That's the obvious case but items are bound to grid, so I have to change
accordingly so removed items don't show in that grid, I'll try to fix that
in latest source code and lets see how it goes.

**Comment by tonysneed** [comment]
One thing to keep in mind is that ApplyChanges doesn't actually persist the changes.  That only takes place when SaveChanges is called.  You might want to think about calling ApplyChanges twice, the first time excluding the deletes, and the second time including only the deletes, which you can resurrect by calling GetChanges().Where(e => e.TrackingState == TrackingState.Deleted).  After both calls to ApplyChanges, you can call SaveChanges, which will wrap all of them in a single transaction.

**Comment by tonysneed** [comment]
@harrisunderwork, ok to close this issue?

**Comment by harrisunderwork** [comment]
Yes sure, please close it.

**Comment by harrisunderwork** [comment]
CollectionDataContract has its own problems with DataContractSerializer.

So I have created DeletedEntities collection for each navigation property using t4 templates. Once serialization is happening, I am moving items from _deletedentities to this newly created collection. On the other end, I am moving back to deletedentities. Rest is taken care by TE.

Thanks.

**Comment by tonysneed** [comment]
Yes, the DCS is deficient regarding serialization hooks for collections -- and it's slow too. Which is why most folks are moving to Json.Net, which is faster, much more flexible, and MS is invested in making sure it stays updated.

**Comment by harrisunderwork** [comment]
Do you have a sample where you have used Json.net with WCF ?

**Comment by tonysneed** [comment]
This is going to depend on your bindings and whether you are using any SOAP headers, or just basic http binding.  If it's just basic, you can add a WebHttpBinding for REST endpoints, which will allow you to use JSON as the wire format.  Here is an [article](http://blogs.msdn.com/b/carlosfigueira/archive/2011/05/03/wcf-extensibility-message-formatters.aspx) which shows show to do that.

**Comment by harrisunderwork** [comment]
Nopes, we are using WSHttpBinding, basicHttpBinding is just too basic :smile:

**Comment by tonysneed** [comment]
Just curious -- are you using a router and need message-level security?

**Comment by harrisunderwork** [comment]
Yup, I am indeed using router + message level security.

**Comment by tonysneed** [comment]
I see why perf is an issue. :smile: So how about instead of removing items from the collection, you just mark them as Deleted?

**Comment by harrisunderwork** [comment]
Sorry Tony, I think you misunderstood. I was talking about performance of ApplyChanges() method.

**Comment by tonysneed** [comment]
Yes, and that was with another issue too. Just :sleepy:

**Comment by harrisunderwork** [comment]
I'll see what I can do to improve the code. I think some parts can be refactored for performance

**Comment by harrisunderwork** [comment]
Can we do parallelism somewhere, I mean use divide and conquer approach while walking the graph ?

**Comment by tonysneed** [comment]
Have you done any profiling yet?  Are you talking just about ApplyChanges, or also SaveChanges?  About how many objects are we talking about?

**Comment by harrisunderwork** [comment]
Yes, I am profiling and found very strange results.

Let me describe here.

Class Contract contains List<ContractLimitGroups> and List<ContractLloydsMarketExtensions>.

ContractLimitGroups contains List<ContractLimits>

1 instance of contract as 2 instance of ContractLimitGroups each of which further contains 3 limits and 1 instance of contract has 1 ContractLloydsMarketExtensions.

Contract => Modifed, ContractLimitGroups, deleted and added and ContractLloydsMarketExtensions modified.

Here is how state is changing for each entity :

4 ContractLimitGroups->Added
4 ContractLimits->Added
4 ContractLimits->Added
4 ContractLimits->Added
6 ContractLloydsMarketExtensions->Modified
4 ContractParticipants->Added
5 ContractMain->Unchanged
4 ContractLimits->Added
4 ContractLimits->Added
4 ContractLimits->Added
6 ContractLimitGroups->Added
6 ContractLimits->Added
4 ContractLimits->Added
6 ContractLimits->Added
6 ContractLimits->Added
6 ContractLimits->Added
6 ContractLimits->Added
4 ContractLimits->Added
6 ContractLimits->Added
6 ContractLimits->Added
6 ContractLimits->Added
6 ContractLimitGroups->Deleted
3 ContractLimits->Unchanged
6 ContractLimits->Deleted
6 ContractLimits->Deleted
6 ContractLimits->Deleted
6 ContractLloydsMarketExtensions->Modified
6 ContractParticipants->Added
6 MasterDataStampUnit->Unchanged
6 ContractParticipants->Deleted
4 ContractLimits->Added
4 ContractLimits->Added
4 ContractLimits->Added
6 ContractLimitGroups->Added
6 ContractLimits->Added
4 ContractLimits->Added
6 ContractLimits->Added
6 ContractLimits->Added
6 ContractLimits->Added
6 ContractLimits->Added
4 ContractLimits->Added
6 ContractLimits->Added
6 ContractLimits->Added
6 ContractLimits->Added
5 ContractLloydsMarketExtensions->Unchanged
6 MasterDataStampUnit->Unchanged
6 ContractParticipants->Added
6 MasterDataStampUnit->Unchanged
6 MasterDataStampUnit->Unchanged

I dont understand if state of entity is already changed, why it keeps setting it again and again.

Second thing is, shouldn't delete order be in reverse ?

**Comment by tonysneed** [comment]
Quick question ... Is the state here *before* SaveChanges is called, or *after*?  And related, are you inspecting EF EntityState via DbContext?  Or are you inspecting items in a database?  And, is this from SQL Profiler?  Or EF's built-in logger?

I notice lot of Adds in there.  Remember that EF is doing its own recursion every time EntityState is set to Added in a graph.  EF actually walks down the graph and sets **everything** to Added.  TE will then need to go and "correct" EF by setting certain items back to Unchanged or Modified.

One more thing, are you using the latest CI NuGet packages from MyGet?  Those include a bug fix for Adds which remove some redundancy in TE.

**Comment by tonysneed** [comment]
PS, My answer to your question on **deletes** will depend on how you're doing the profiling.

**Comment by harrisunderwork** [comment]
1) This state is before calling SaveChanges, all this logging is done in DBContextExtensions file. 
2) I am inspecting EntityState. This is from my custom Debug statements. I have them where you are changing state of items.

Regarding changing item to unchanged to set only modified properties to true is perfectly okay. 

My solution is directly attached to source code so I am not taking any binaries.

**Comment by harrisunderwork** [comment]
Okay, I think I know why this issue is there, ContractMain too has reference of ContractLimits.

**Comment by tonysneed** [comment]
OK, so make sure you are on the dev branch, and you should have the latest fix in there for #51 with commit be7041c47e4f479655342c603e7de7a30c9a36f1.  One thing I noticed is that, on [line 131](https://github.com/tonysneed/trackable-entities/blob/be7041c47e4f479655342c603e7de7a30c9a36f1/Source/TrackableEntities.EF.5/DbContextExtensions.cs#L131), we could add a condition that checks the EntityState and only sets it to Added if it's not already Added.  That would save some cycles, because when EF sets EntityState to Added, it does a graph traversal.

To answer your question on **deletes**:
> Shouldn't delete order be in reverse?

The order in which EntityState is set to Deleted is irrelevant.  Actually, EF dictates the order in which changes to EntityState need to be applied.  It is only when ```SaveChanges``` is called that the order of inserts, updates and deletes matter, and EF controls that according to ref integrity rules.

In terms of setting EntityState more than once on the same entity, I'd have to see how you are looking at that.  *The object visitor should prevent that from happening.*

**Comment by harrisunderwork** [comment]
Took the change few minutes ago.

I think delete is making a mess here as the entities being deleted are having two parents. They are not being prevented by object visitor as at some places we are using their clones I guess.

**Comment by tonysneed** [comment]
> I think delete is making a mess here as the entities being deleted are having two parents. They are not being prevented by object visitor as at some places we are using their clones I guess.

The way deletes work in TE is that the children are marked for deletion before the parents.  The best way to tell what is going on there is to create a small, focused unit test to see what happens when a child has two parents and to make sure the object visitor is doing its job.

I think I asked this before, but instead of removing entities in order to delete them, which clones and caches them, *why don't you instead simply* **set TrackingState to Deleted** *on the entities you wish to delete?*  Then there's no need to re-add them or worry about serializing the clones, etc.

AFAIK, EF only recursively sets state to Added (not Modified or Deleted).  Cascading deletes are done at the provider level and depend on how the model relations are defined.

**Comment by harrisunderwork** [comment]
Problem is, STE had MarkAdModified extension method. Primary function of that method was to mark entity deleted and remove from collection. Now lot of business logic is built on it (deleting a row and prevent creation of new row with same primary key). Now, if just tracking state,

1) UI controls need to be changed to handle TrackingState which I don't think is good from separation of concern point of view.
2) Business logic will have information of TrackingState too.

I'll try to create a small unit test case and make a pull request.

**Comment by tonysneed** [comment]
Re: setting TrackingState to Deleted, I see now what your situation is.

Looking forward to your PR! :+1:

**Comment by harrisunderwork** [comment]
To create a new branch and publish it, do I need rights from your side ?

**Comment by tonysneed** [comment]
Nope, you first fork the TE repo. Then clone to get a local copy of your fork. Then creat a local branch from dev. Use VS 2013, which has Git integration. Also TortoiseGit is good.

**Comment by harrisunderwork** [comment]
Hi Tony, how deletion of entites work in EF6 ? Lets say I mark a parent with some child deleted, will it delete children too ? What happens when that child has multiple parent ?

**Comment by tonysneed** [comment]
I lot depends on how you have configured the entity relations.  If you have not specified cascade delete (which you would need to do from both parents), then you always must delete the children before deleting the parent(s).  Otherwise, EF will throw an exception.

You can see this in action for the DeleteOrder method of the WcfSample, where children must be loaded (via the Include operator).  Internally, ApplyChanges deletes the children first.

So to answer your question, it doesn't really matter that a child has two parents, so long as you first mark the children as deleted (usually via a decrementing for loop), because marking either parent as deleted.

**Comment by harrisunderwork** [comment]
Got it, started reading Programming Entity Framework : DbContext.

I guess there is some issue with hierarchical model with TE. Though it is
marking all of them deleted, but the order of deletion is wrong. It is
trying to delete Parent first and then children. Is this something EF
overrides while executing the queries ? I checked the order in
ObjectStateManager.

**Comment by tonysneed** [comment]
For the delete to work properly, the children must be eager loaded by a query that uses the Include operator.  The reason you may see the parent being deleted first is that the children are not loaded (just a guess on my part).

Do you have cascade delete defined on the relations from both parents to the common children?

**Comment by harrisunderwork** [comment]
Cascade delete is set false and thats what i guess is wrong on my part.
Though children are present too.

Do you know how cascade delete works when using database first approach ?

Is it something we need to configure explicitly ? I have used reverse code
first approach to generate POCO classes from database design in my test
case.

Also, do the test cases needs to have generalized classes ? I have used
same names as in the business application. Though I think to rename them
when I perform PR.

**Comment by tonysneed** [comment]
If cascade delete is set to false, then you definitely need to have the deletes take place first on the children before either parent is marked for deletion.  I haven't tested the scenario of deleting entities with two parents, so it's possible TE does not handle this.  But I would need to see a test case to be certain.

You could try configuring cascading deletes, but you would do this on the relations in the database first, then reverse engineering the entities should pick this up and represent it in the model.

Have you tried performing the deletes manually without TE?

For the unit tests, you can create any kind of entities you wish, but I do suggest you create general purpose classes rather than the same as your business app, so that they can be as simple as possible.

**Comment by harrisunderwork** [comment]
Sure, let me first confirm everything on my side and then submit pr.

I have tried manually by just deleting parent, it was throwing exception.

**Comment by tonysneed** [comment]
You have to try manually by deleting children before parents.

**Comment by harrisunderwork** [comment]
Hi Tony,

I have tracked down the issue. When the parent has a child deleted and same
type of parent is added, it is performing change state of child to added,
therefore entitystate of all objects, including deleted one are also
converting to added. Next time when deleted needs to be processed, its
entitystate is added and target is deleted, it is just failing to delete
any relation associated with it and thats causing the mess.

To fix that, I just made a generic AppyChanges in first I attach the parent
object to context and then ApplyChanges().

It fixed the problem :)

**Comment by harrisunderwork** [comment]
Hi Tony,
Because of my fix, many of test cases were failing. So I m trying to fix in TE itself. Problem is when you are changing state of any child entity to Added, it is changing entitystate of all entities(deleted or modified) to added too. So next time when a deleted child entity is changing state (Added -> Deleted), it fails to delete the relationship which the child entity is having with its own child entity. To fix the problem so test cases don't fail, I am trying to check when entitystate of a child entity is being changed to added, I should also check whether parent is attached or not. If parent is not attached, it should attach parent to context first. 

Now my test case is passing with that, but in many of your test cases, primary key assigned as 0, that is preventing the attach of parent to happen.

Do you think it was a good idea to check parent first ? Or do you think otherwise that its okay to assign negative integers to newly added entries ?

**Comment by tonysneed** [comment]
According to EF rules, when a parent in a 1-M relationship is set to Added, then the children are also set to Added.  This is because the assumption from EF's side is that children cannot be orphans, that is, a child in a 1-M relation must have a parent.  And TE respects this rule.

All of this works well if you are only calling ```ApplyChanges``` **once** prior to calling ```SaveChanges```.  If you are calling ApplyChanges *more than once*, then you're working against some basic assumptions which both EF and TE make.  As I'm unfamiliar with the fixup code, I'm uncertain as to why tracking state is being affected at all.  In fact, I think I advised to turn off change tracing altogether while the fixup is taking place.

EF will not complain about PK values of zero while EntityState is set to Added.  So again, I think your troubles may all stem from *not disabling change tracking* while fixing up the relationships: ```ChangeTrackingCollection<T>.Tracking = false```.  (Collections in TE are all of this type.)

**Comment by harrisunderwork** [comment]
Well, I have reverted the source code of TE and added one more method,
ClearFixupValues.

The main function of this method is to nullify M-1 entities as to clear the
fixup values. I am still concerned about valid M-1 entities that are not of
inverse relation. Your TE works as expected, even EF is working as
expected, problem is, starting with EF 6, they have decided not to pay attn
to fixup values, but since I used them in my solution, it started getting
messy.

Do you know any way to determine inverse properties so that I can clear
them up before ApplyChanges() ?

**Comment by tonysneed** [comment]
Because I'm unfamiliar with the fixup code, I'm not in a position to advise on how to reverse it.  But I imagine that you would need some way to determine which properties represent fixed up relations, so that you can then set those to null.

Interestingly, I'm not 100% certain, but I believe EF itself fixes up the relations when calling ```SaveChanges```.

**Comment by harrisunderwork** [comment]
Yes, you are right, it does fixes up the relations but only while calling
SaveChanges(), not anytime before that.

http://stackoverflow.com/questions/6723784/should-i-rip-out-the-association-fixup-code-from-my-entity-framework-t4-template
, 2nd answer.

**Comment by tonysneed** [comment]
:+1: on removing fixup code from T4 templates. :smile:  I understand, however, that your business logic depends on it.  But what I would do is perform relationship fixup with code which is external to the entities, in which you explicitly disable change tracking before fixup, then restore change tracking to true if it was turned on prior to the fixup.

**Comment by harrisunderwork** [comment]
That is already done. I have enabled change tracking once fixup has
happened. By fixup I mean, a child entity is related to parent entity by id
and a navigation property. I think in all of your examples, you haven't
specified inverse navigation property. Once you specify it, irrational
behavior is bound to happen.

This is the method which I have called to clear fixup values.

https://gist.github.com/harrisunderwork/da3fb5afee5d802eae30

Once this is done, entity state changes only flow from parent to child, not
reverse and everything works as expected.

I also have added check not to change entity state to added if it is
already added to prevent extra processing.

**Comment by harrisunderwork** [comment]
Closing this bug for now.

ClearFixUpValues method worked for me. Fixup is NOT recommended unless you have strong reasons to do so. I also added some checks where EntityState is being changed to Added, which removed extra iterations.
